src/main/java/edu/kit/datamanager/security/filter/KeycloakTokenValidator.java

```diff
@@ -0,0 +1,237 @@
+     * Must not be negative.
+     */
+    private int connectTimeoutms = 0;
+    private int readTimeoutms = 0;
+    private int sizeLimit = 0;
+    private boolean initialized = false;
+    private String jwtLocalSecret = null;
+
+    public void setJwtProcessor(ConfigurableJWTProcessor jwtProcessor) {
+        this.jwtProcessor = jwtProcessor;
+    }
+
+    private ConfigurableJWTProcessor jwtProcessor;
+
+    private JWSKeySelector keySelector(JWKSource keySource) {
+        return new JWSVerificationKeySelector(JWSAlgorithm.RS256, keySource);
+    }
+
+    private void init(String jwkUrl) {
+        if (jwkUrl != null) {
+            LOG.info("Initializing JWK set from {}.", jwkUrl);
+            try {
+                JWKSet jwkSet = getJwkSet(jwkUrl);
+                JWKSource keySource = new ImmutableJWKSet(jwkSet);
+                jwtProcessor.setJWSKeySelector(keySelector(keySource));
+                LOG.info("JWK set initialized successfully.");
+                initialized = true;
+            } catch (IOException | ParseException e) {
+                throw new RuntimeException("Failed to initialize KeycloakTokenValidator.", e);
+            }
+        }
+    }
+
+    public JWKSet getJwkSet(String jwkUrl) throws IOException, ParseException {
+        return JWKSet.load(new URL(jwkUrl), connectTimeoutms, readTimeoutms, sizeLimit);
+    }
+
+    public JwtAuthenticationToken validate(String accessToken) throws BadJOSEException {
+        SecurityContext ctx = null;
+        try {
+            JWTClaimsSet claimsSet = getJwtClaimsSet(accessToken, ctx);
+            if (claimsSet != null) {
+
+                List<String> aud = claimsSet.getAudience();
+                if (aud == null || !aud.contains(resource)) {
+                    throw new BadJWTException("Invalid Keycloak Resource. Audience claim 'aud' is missing.");
+                }
+
+                //build auth token
+                JSONArray aRoles = null;
+
+                JSONObject o = claimsSet.getJSONObjectClaim("realm_access");
+                if (o != null) {
+                    aRoles = (JSONArray) o.get("roles");
+
+                }
+
+                if (aRoles == null) {
+                    aRoles = new JSONArray();
+                    aRoles.add("GUEST");
+                }
+
+                //token type? Service, Temp, User?
+                Map<String, Object> claims = new HashMap<>();
+                String roles = aRoles.toJSONString();
+                claims.put("username", claimsSet.getStringClaim((jwtClaim == null) ? "preferred_user" : jwtClaim));
+                claims.put("firstname", claimsSet.getStringClaim("given_name"));
+                claims.put("lastname", claimsSet.getStringClaim("family_name"));
+                claims.put("email", claimsSet.getStringClaim("email"));
+                claims.put("roles", roles);
+                return JwtAuthenticationToken.factoryToken(accessToken, claims);
```

Comment:
JwtAuthenticationToken returnValue = null;
returnValue = JwtAuthenticationToken.factoryToken(accessToken, claims);
return returnValue;

Reply:
Hm, why not...

Reply:
public JwtAuthenticationToken validate(String accessToken) throws BadJOSEException {
      JwtAuthenticationToken returnValue = null;
[...]
returnValue = JwtAuthenticationToken.factoryToken(accessToken, claims); (instead of return)
[...]
return returnValue;

Sorry, that's a spleen of mine that the return should only be at the end. I don't like it when I have to search all returns in between for long methods.